### PR TITLE
Xenial   import vpn files

### PR DESCRIPTION
**Pull request by Pekkari**

### CODE REVIEW

**Comment by Pekkari** [comment]
@Flohack74 here it is the new PR with the branch corrected.

Thanks!

**Comment by UniversalSuperBox** [comment]
Does this depend on any other pull requests?

**Comment by Pekkari** [comment]
No but, [this](https://gitlab.com/ubports/core/lomiri-system-settings/-/merge_requests/272) depends on this PR

File: debian/control [hunk 1]
```diff
@@ -48,8 +49,10 @@ Multi-Arch: foreign
 Depends: ${misc:Depends},
          ${shlibs:Depends},
          indicator-common,
+         libnm0,
```
Comment (dobey) [inline debian/control]: This shouldn't need to be here, as it should be added automatically via the `${shlibs:Depends}` above.

File: debian/control [hunk 2]
```diff
@@ -48,8 +49,10 @@ Multi-Arch: foreign
...
          network-manager,
          network-manager-openvpn,
+         network-manager-openvpn-gnome,
```
Comment (dobey) [inline debian/control]: I suppose this is necessary due to the version of NM we have currently in 16.04? What all else does this cause to get pulled in?

File: src/indicator/CMakeLists.txt [hunk 3]
```diff
@@ -143,6 +143,7 @@ target_link_libraries(
     ${QOFONO_LDFLAGS}
     Qt5::Core
     Qt5::DBus
+    nm
```
Comment (dobey) [inline src/indicator/CMakeLists.txt]: This should be `${NM_LIBRARIES}` instead.

File: src/indicator/CMakeLists.txt [hunk 4]
```diff
@@ -159,6 +160,7 @@ target_link_libraries(
     indicator-network-service-static
     Qt5::Core
     Qt5::DBus
+    nm
```
Comment (dobey) [inline src/indicator/CMakeLists.txt]: Same as above for `${NM_LIBRARIES}`.

File: src/indicator/nmofono/vpn/vpn-manager.cpp [hunk 5]
```diff
@@ -278,6 +279,67 @@ QString VpnManager::addConnection(VpnConnection::Type type)
     return uuid;
 }
 
+QString VpnManager::importConnection(QString filepath)
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-manager.cpp]: It seems like this should be changed to accept the `type` as an argument as well, in case we want to support importing config for other VPN types in the future. We can also use that information to load a specific plug-in to do the import with, rather than simply looping through all the plug-ins and attempting to import in every plug-in as is done here with the copy and paste code from the upstream example.

File: src/indicator/nmofono/vpn/vpn-manager.cpp [hunk 6]
```diff
@@ -278,6 +279,67 @@ QString VpnManager::addConnection(VpnConnection::Type type)
...
+        g_print("Import VPN connection error="
+            "didn't get a filename back from the chooser!"
+        );
+        throw domain_error(qPrintable("Import VPN connection failed: "
+            "didn't get a filename back from the chooser!"));
+    }
+
+    plugins = nm_vpn_plugin_info_list_load();
+
+    for (iter = plugins; iter; iter = iter->next) {
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-manager.cpp]: Instead of looping through a list, we should have a mapping of supported plugins for the enumerated integer type used internally, to the plugin names used by libnm. With that, we can use `nm_vpn_plugin_info_list_find_by_name()` to get the specific plug-in we're importing for. This would greatly simplify the code in this function, and make it more appropriate for how the rest of the connectivity service here works.

File: src/indicator/nmofono/vpn/vpn-manager.cpp [hunk 7]
```diff
@@ -278,6 +279,67 @@ QString VpnManager::addConnection(VpnConnection::Type type)
     return uuid;
 }
 
+QString VpnManager::importConnection(QString filepath)
+{
+    NMConnection *conn = NULL;
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-manager.cpp]: Please initialize using `nullptr` insetad of `NULL` when using in C++, even when using C APIs.

File: src/indicator/nmofono/vpn/vpn-manager.cpp [hunk 8]
```diff
@@ -278,6 +279,67 @@ QString VpnManager::addConnection(VpnConnection::Type type)
...
+        break;
+    }
+    g_slist_free_full(plugins, g_object_unref);
+
+    if (!conn) {
+        g_print("Failure to import the file with any plugin\n");
+        return NULL;
+    }
+
+    return strcat("imported connection: ", nm_connection_get_id(conn));
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-manager.cpp]: This creates a memory leak as `strcat()` returns a newly allocated `char*`, and `conn` object is not being freed. Also, it seems the code above which is calling this function expects a string that is a UUID only. Therefore, it might be better to use `QUuid` as the return type for this `importConnection` method,, and create a new variable such as `auto uuid = QUuid::fromString(nm_connection_get_id(conn));`, then `g_clear_object(&conn);` and finally do `return uuid;` here.

This will avoid those leaks, and provide a strong type for the expected result.

**Review by dobey — CHANGES_REQUESTED** [review]
Thanks for working on this! It's a good start, and a much better approach than the previous attempt. I've spotted a few small problems that I've commented on inline though. :)

**Comment by Pekkari** [comment]
Thanks for the comments! Very insightful! I'm a bit busy, so please bear with it. I'll be taking time to address them tomorrow night. I'll come back.

File: debian/control (hunk 2 above)
Comment (Pekkari) [reply debian/control]: this seems to have become redundant, at the time of doing the patch was bringing:

/usr/lib/aarch64-linux-gnu/NetworkManager/libnm-vpn-plugin-openvpn.so

Which is needed for the code, but now I see this is living in the network-manager-openvpn,
where I expected to find it at first.

File: src/indicator/nmofono/vpn/vpn-manager.cpp (hunk 5 above)
Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: I personally see no problem in looping in a list of 5-6 plugins and try them all, but it is a valid request, I'll take it up

File: debian/control (hunk 1 above)
Comment (Pekkari) [reply debian/control]: this doesn't seem to happen, multiple references inside the function are no longer found if I drop it.

Comment (Pekkari) [reply debian/control]: Actually it seems to be the CMakeLists.txt change the one producing the miss

File: src/indicator/CMakeLists.txt (hunk 4 above)
Comment (Pekkari) [reply src/indicator/CMakeLists.txt]: I think this is LIBNM_LIBRARIES instead, since otherwise it doesn't bring the package I'm looking for

File: src/indicator/nmofono/vpn/vpn-manager.cpp (hunk 8 above)
Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: it seems for this function and the similar like addVpn, etc, they do the calls and take results by using QString, so, if you don't mind I'll follow the example and return a QString here, and it's up to a further change to move all the function prototypes to return QUuids instead.

**Comment by Pekkari** [comment]
Alright, new version ready for review. Thanks again @dobey !

File: src/indicator/nmofono/vpn/vpn-connection.h [hunk 9]
```diff
@@ -46,7 +46,13 @@ class VpnConnection: public QObject
...
     {
         openvpn,
-        pptp
+        pptp,
+        fortisslvpn,
+        iodine,
+        l2tp,
+        openconnect,
+        strongswan,
+        vpnc
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-connection.h]: Is this change necessary, given none of these other types are supported? It seems like we should only add them when we include support for them, no?

File: src/indicator/nmofono/vpn/vpn-manager.cpp [hunk 10]
```diff
@@ -278,6 +280,63 @@ QString VpnManager::addConnection(VpnConnection::Type type)
     return uuid;
 }
 
+QString VpnManager::importConnection(QString filepath, VpnConnection::Type type)
```
Comment (dobey) [inline src/indicator/nmofono/vpn/vpn-manager.cpp]: Please adjust the added code below to match the coding style of the rest of the indicator. I think the `filepath` argument should also be taken as `const QString& filepath` instead.  It would probably also be good for the argument order to match that of the function which is calling into this one, for consistency.

As another point, all of the `g_print()` statements in this method should be removed. Many also don't make so much sense any more, as we're not looping through all plugins, and are mostly extraneous things only useful for the example code. It looks like there might still be a couple memory leaks, and it might be a good idea to wrap some of the libnm types as a `std::shared_ptr` to improve the memory management and make it easier to just `throw runtime_error();` in the cases where errors occur. That would simplify the code, and make it more suitable for future modification if needed, as well.

**Review by dobey — CHANGES_REQUESTED** [review]
A few more points, but it's definitely looking better and coming together now.

Thanks again for working on this!

File: src/indicator/nmofono/vpn/vpn-connection.h (hunk 9 above)
Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-connection.h]: yep, that is probably the way to go, I'll change this to pptp and openvpn

File: src/indicator/nmofono/vpn/vpn-manager.cpp (hunk 10 above)
Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: Alright, I'm moving the filepath type, though I didn't find the order match you are looking for, the function calling is sending filepath first, and the vpn connection type afterwards, see below:
```
QString uuid = p.d->m_vpnManager->importConnection(
                    filepath,
                    static_cast<VpnConnection::Type>(type)
            );
```

Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: aw, I see it now, it is the function parameters of the calling function:
```
QDBusObjectPath PrivateService::ImportVpnConnection(int type, const QString &filepath) {
[...]
```

Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: The new version addresses these comments, except for the std::shared_ptr/throw runtime_error() comment for now. I'll take a look what I can do in that direction.

Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: Hi, sorry, it seems that we can't do the std::shared_ptr since the NMConnection and NMVpnEditorPlugin are incomplete types enforcing the necessity of having a destructor, see the output below:
```
In file included from /usr/include/c++/5/bits/shared_ptr.h:52:0,
                 from /usr/include/c++/5/memory:82,
                 from /usr/include/unity/util/DefinesPtrs.h:22,
                 from /home/phablet/indicator-network/src/indicator/nmofono/connection/active-connection.h:25,
                 from /home/phablet/indicator-network/src/indicator/nmofono/connection/active-connection-manager.h:24,
                 from /home/phablet/indicator-network/src/indicator/nmofono/vpn/vpn-manager.h:22,
                 from /home/phablet/indicator-network/src/indicator/nmofono/vpn/vpn-manager.cpp:21:
/usr/include/c++/5/bits/shared_ptr_base.h: In instantiation of ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(_Tp1*) [with _Tp1 = _NMVpnEditorPlugin; _Tp = _NMVpnEditorPlugin; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’:
/usr/include/c++/5/bits/shared_ptr.h:117:32:   required from ‘std::shared_ptr<_Tp>::shared_ptr(_Tp1*) [with _Tp1 = _NMVpnEditorPlugin; _Tp = _NMVpnEditorPlugin]’
/home/phablet/indicator-network/src/indicator/nmofono/vpn/vpn-manager.cpp:305:5:   required from here
/usr/include/c++/5/bits/shared_ptr_base.h:886:25: error: invalid application of ‘sizeof’ to incomplete type ‘_NMVpnEditorPlugin’
    static_assert( sizeof(_Tp1) > 0, "incomplete type" );
                         ^
/usr/include/c++/5/bits/shared_ptr_base.h: In instantiation of ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(_Tp1*) [with _Tp1 = _NMConnection; _Tp = _NMConnection; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’:
/usr/include/c++/5/bits/shared_ptr.h:117:32:   required from ‘std::shared_ptr<_Tp>::shared_ptr(_Tp1*) [with _Tp1 = _NMConnection; _Tp = _NMConnection]’
/home/phablet/indicator-network/src/indicator/nmofono/vpn/vpn-manager.cpp:315:5:   required from here
/usr/include/c++/5/bits/shared_ptr_base.h:886:25: error: invalid application of ‘sizeof’ to incomplete type ‘_NMConnection’
src/indicator/CMakeFiles/indicator-network-service-static.dir/build.make:598: recipe for target 'src/indicator/CMakeFiles/indicator-network-service-static.dir/nmofono/vpn/vpn-manager.cpp.o' failed
```

Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: I pushed now a last update to my patch which at least wraps the NMVpnPluginInfo type, which seems
to be the only one that is not affected by an incomplete type. For now, it seems the best we can do. Let
me know if you find it otherwise.

Comment (Pekkari) [reply src/indicator/nmofono/vpn/vpn-manager.cpp]: Hi @dobey, can I assume this is fine and we are good to go with the patch? Is there any more comments/
improvements you'd like me to pursue?

Thanks for the comments!

José.

**Review by dobey — CHANGES_REQUESTED** [review]
OK. I think there was some miscommunication, or misunderstanding, so I made a PR into your branch with the cleanups I was suggesting, which should ensure the memory is not leaked, and errors get propagated properly.

After you approve and merge that in, we need some regression and integration tests added, to ensure things don't break in the future and everything keeps working correctly if/when new VPN types are added and such.

Thanks!

**Comment by Pekkari** [comment]
Alright, merged, thanks for the patch!

**Comment by Pekkari** [comment]
The test is added, though I fail to find what is wrong right now, if it is the test being wrong, or the code is broken at this point, could I borrow some eyes to sort the thing out?

Thanks!

File: tests/integration/indicator-network-test-base.cpp [hunk 11]
```diff
@@ -604,6 +605,29 @@ QString IndicatorNetworkTestBase::createVpnConnection(const QString& id,
...
+    const QString& ovpn_sample("sample.ovpn");
+
+    auto reply = vpnManager->importConnection(
+        nmofono::vpn::VpnConnection::Type::openvpn,
+        ovpn_sample
+    );
+    if (reply.isNull())
+    {
+        EXPECT_FALSE("Connection import failed");
+    }
```
Comment (dobey) [inline tests/integration/indicator-network-test-base.cpp:627]: This block doesn't really make any sense. `EXPECT_FALSE()` expects a boolean expression. So what are you trying to do here? I guess there's also no need to check reply here, as it's still being returned anyway, and the test won't stop processing.

File: tests/integration/indicator-network-test-base.cpp [hunk 12]
```diff
@@ -604,6 +605,29 @@ QString IndicatorNetworkTestBase::createVpnConnection(const QString& id,
...
+    nmofono::vpn::VpnManager::SPtr vpnManager = make_shared<nmofono::vpn::VpnManager>(
+        actConnManager,
+        QDBusConnection::systemBus()
+    );
+    const QString& ovpn_sample("sample.ovpn");
+
+    auto reply = vpnManager->importConnection(
+        nmofono::vpn::VpnConnection::Type::openvpn,
+        ovpn_sample
+    );
```
Comment (dobey) [inline tests/integration/indicator-network-test-base.cpp:623]: For this to work, I think that `network-manager-openvpn` also needs to be added to `Build-Depends` in `debian/control` otherwise the plugin will fail to load.

**Review by dobey — CHANGES_REQUESTED** [review]

**Comment by dobey** [comment]
> The test is added, though I fail to find what is wrong right now, if it is the test being wrong, or the code is broken at this point, could I borrow some eyes to sort the thing out?
> 
> Thanks!

The test failed because the plugin couldn't be loaded, and so it threw an exception.

File: tests/integration/indicator-network-test-base.cpp (hunk 11 above)
Comment (Pekkari) [reply tests/integration/indicator-network-test-base.cpp:627]: My bad, I copied wrongly from the examples around, testing is always a bit daunting to me, so these kind of oddities comes out of my mind

File: tests/integration/indicator-network-test-base.cpp (hunk 12 above)
Comment (Pekkari) [reply tests/integration/indicator-network-test-base.cpp:623]: adding, thanks!

Comment (Pekkari) [reply tests/integration/indicator-network-test-base.cpp:623]: > The test failed because the plugin couldn't be loaded, and so it threw an exception.

Thanks for the hint, though I'm afraid there is something else, since after loading the plugin and running the code, the test expects to find one element that is the new connection, and another which is the config wheel to modify it. These elements doesn't turn up.

**Comment by Pekkari** [comment]
This is odd, I managed to fix the code and the test case, and I see it perfectly fine in the ci, though it
reports another test case failed while there is no other case that should take this codebase. Also
I fail to see the failed test. Can I have some other eyes around?
```
[ RUN      ] TestIndicatorVpn.importConnection

Debug: /build/indicator-network-0.8.1+ubports1+0~20210912130614.13~1.gbpfae40b/tests/data/networkmanager.py ((null):0, (null))

1631452251.607 GetAll org.ofono.Modem

1631452251.608 emit org.ofono.Manager.ModemAdded "/ril_0" {"Powered": True, "Manufacturer": "Fakesys", "Serial": "12345678-1234-1234-1234-000000000000", "Lockdown": False, "Online": True, "Features": ["gprs", "net"], "Interfaces": ["org.ofono.CallVolume", "org.ofono.VoiceCallManager", "org.ofono.NetworkRegistration", "org.ofono.SimManager", "org.ofono.ConnectionManager"], "Revision": "0815.42", "Emergency": False, "Model": "Mock Modem", "Type": "hardware"}
[...]
1631452251.819 emit org.freedesktop.DBus.Properties.PropertiesChanged "org.ofono.ConnectionManager" {"RoamingAllowed": False} []

1631452251.819 emit org.ofono.ConnectionManager.PropertyChanged "RoamingAllowed" False

[       OK ] TestIndicatorVpn.importConnection (847 ms)

[----------] 5 tests from TestIndicatorVpn (4130 ms total)
```

**Comment by Pekkari** [comment]
This seems to be hitting a bug in ctest, where all the integration tests actually passes if you run
them one by one, but when you ask ctest to run them all, it reaches the TestConnectivityApi and
crashes right away:
```
[ RUN      ] TestConnectivityApi.OnlineStatus
Debug: /home/phablet/indicator-network/tests/data/networkmanager.py ((null):0, (null))


/build/cmake-WiKIfW/cmake-3.5.1/Source/cmCTest.cxx:156    Current_Time: Sep 12 19:05 EEST

/build/cmake-WiKIfW/cmake-3.5.1/Source/CTest/cmCTestMultiProcessHandler.cxx:127 test 2

/build/cmake-WiKIfW/cmake-3.5.1/Source/CTest/cmCTestRunTest.cxx:447     Start 2: unit-tests
```
Running this way:
```
$ tests/integration/integration-tests --gtest_filter=TestConnectivityApi.*
[...]
[       OK ] TestConnectivityApi.SettingsRestoredOnStartup (5105 ms)
[----------] 11 tests from TestConnectivityApi (55866 ms total)

[----------] Global test environment tear-down
[==========] 11 tests from 1 test case ran. (55867 ms total)
[  PASSED  ] 11 tests.
```

**Comment by Pekkari** [comment]
by changing the order of execution in the test to reproduce the crash and running
under gdb I find this trace:
```
Debug: /home/phablet/indicator-network/tests/data/networkmanager.py ((null):0, (null))

Thread 1 "integration-tes" received signal SIGTERM, Terminated.
0x0000007fb710bd60 in raise (sig=15)
    at ../sysdeps/unix/sysv/linux/pt-raise.c:35
35      ../sysdeps/unix/sysv/linux/pt-raise.c: No such file or directory.
(gdb) bt
#0  0x0000007fb710bd60 in raise (sig=15)
    at ../sysdeps/unix/sysv/linux/pt-raise.c:35
#1  0x0000007fb69b7e60 in ffi_call_SYSV ()
   from /usr/lib/aarch64-linux-gnu/libffi.so.6
#2  0x0000007fb69b87b8 in ffi_call ()
   from /usr/lib/aarch64-linux-gnu/libffi.so.6
#3  0x0000007fb7700368 in g_cclosure_marshal_generic_va ()
   from /usr/lib/aarch64-linux-gnu/libgobject-2.0.so.0
#4  0x0000007fb76fdb78 in ?? ()
   from /usr/lib/aarch64-linux-gnu/libgobject-2.0.so.0
#5  0x0000007fffffe130 in ?? ()
#6  0x0000000000000006 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

**Comment by peat-psuwit** [comment]
Hello,

Thank you for contributing to UBports. As part of project renaming and the effort to port Ubuntu Touch stack to Ubuntu 20.04, we're incrementally migrating repositories to GitLab.

Your PR is now migrated to:
https://gitlab.com/ubports/core/lomiri-indicator-network/-/merge_requests/68
However, you might not be able to push new code into it, as the branch resides in our repository. In case you need to update it, please open a new MR from your (new) fork. I don't have a better solution at the moment.

Sorry for your inconvenience.
